src/hotspot/share/runtime/handshake.cpp

```diff
@@ -318,6 +319,10 @@ void HandshakeOperation::do_handshake(JavaThread* thread) {
   // Only actually execute the operation for non terminated threads.
   if (!thread->is_terminated()) {
     _handshake_cl->do_thread(thread);
+  } else {
+    if (_handshake_cl->is_async_installer()) {
+      _handshake_cl->do_cleanup();
+    }
```

Comment:
Rather than defining this is_async_installer() I would just add a is_exiting() check in JavaThread::install_async_exception(), like we do with suspend_with_handshake(), so we never install an async exception if the JT already moved to the _thread_exiting state. This !thread->is_terminated() check still succeeds if the thread moved to the _thread_exiting state, so we could still install an async exception between the JT moving to _thread_exiting and moving to _thread_terminated (the JT might block when trying to grab the Threads_lock to remove itself from the thread list). So in that case, we would still leak memory. (Maybe we should just check for _not_terminated here, not sure why we allow executing the handshake closure when thread is already exiting).

To avoid future issues I would define ~HandshakeState() with a single debug line: assert(!has_operation(), "leaking memory, queue should be empty"). If you add the following code to StopAtExit.java this assert will fire do to what I mentioned about the !thread->is_terminated() check still succeeding between _thread_exiting and _thread_terminated:

```
// Fire-up thread that just creates new threads contending on Threads_lock
Thread threadCreator = new Thread() {
    @Override
    public void run() {
        while (true) {
            Thread dummyThread = new Thread(() -> {});
            dummyThread.start();
            try {
                dummyThread.join();
            } catch(InterruptedException ie) {
            }
        }
    }
};
threadCreator.setDaemon(true);
threadCreator.start();
```

Reply:
It appears that adding an `is_exiting()` check to `JavaThread::install_async_exception()`
is insufficient. I restored my debug counter code and disabled the two leak fixes in this
PR and retested:

```
vm_exit: dcubed_async_global_alloc_cnt=936344
vm_exit: dcubed_drop_installer_thread_terminated_cnt=113472
vm_exit: dcubed_drop_is_exiting_cnt=13
vm_exit: dcubed_async_blocked_cnt=67577
dcubed_async_global_release_cnt=755282
```

dcubed_async_global_alloc_cnt=936344 - dcubed_async_global_release_cnt=755282
181062 // The number of exceptions leaked.

181062 - dcubed_drop_is_exiting_cnt=13
181049 // The number of leaked exceptions that made it past the is_exiting() check.

181049 - dcubed_drop_installer_thread_terminated_cnt=113472
67577 // The number of exceptions that made it past the is_exiting() check and
           // the is_terminated() check in HandshakeOperation::do_handshake().

67577 - dcubed_async_blocked_cnt=67577
0   // The number of exception that made it past the cleanup in JavaThread::exit

Reply:
The clean_async_exception_operation() call in exit() shouldn't be removed, only this is_async_installer() part.

Reply:
The target thread sets the `_terminated` field to `_thread_exiting` in
src/hotspot/share/runtime/thread.cpp: JavaThread::exit() AFTER the
creation of the `NoAsyncExceptionDeliveryMark` object that protects
the VM call out to `Thread::exit()`. The dcubed_async_blocked_cnt
counter I mumble about above is counted before the target thread
sets the `_terminated` field to `_thread_exiting` so that's why adding
the new `is_exiting()` check is so ineffective.

Reply:
I don't understand, was the clean_async_exception_operation() call in JavaThread::exit() removed? The only place where another async exception could have been installed after the call to clean_async_exception_operation()(if not removed) and the change of state to _thread_exiting is in post_thread_end(). So if anything we would need to call it after that. Is your testing calling that jvmti event?

Reply:
Sorry for confusing you. I probably shouldn't be mumbling about my debugging
here in the PR. Please ignore the debugging comments above.